### PR TITLE
Ghila reentrancy examples

### PR DESCRIPTION
**Pull request by ghilac**

### CODE REVIEW

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBank.conf [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+    "files": [
+        "contracts/VulnerableBank.sol"
+    ],
+    "loop_iter": "3",
+    "optimistic_loop": true,
+    "verify": "VulnerableBank:certora/spec/Reentrancy.spec",
+    "prover_args": [
+         "-ignoreViewFunctions"
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBank.conf:9]: This is no longer needed.

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "files": [
+        "contracts/VulnerableBankBadFix.sol"
+    ],
+    "loop_iter": "3",
+    "optimistic_loop": true,
+    "verify": "VulnerableBankBadFix:certora/spec/Reentrancy.spec",
+    "prover_args": [
+         "-ignoreViewFunctions"
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf:9]: same

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    ],
+    "prover_version":"shelly/curve",
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf:11]: Is this the right branch? Will it always stay?

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankFixed.conf [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{
+    "files": [
+        "contracts/VulnerableBankFixed.sol"
+    ],
+    "loop_iter": "3",
+    "optimistic_loop": true,
+    "verify": "VulnerableBankFixed:certora/spec/Reentrancy.spec",
+    "prover_args": [
+         "-ignoreViewFunctions"
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankFixed.conf:9]: same

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBank.conf (hunk 1 above)
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBank.conf:9]: But you filter in the spec to non view so why do you need this?

File: CVLByExample/Reentrancy/certora/spec/Reentrancy.spec [hunk 5]
```diff
@@ -0,0 +1,27 @@
+ghost bool called_extcall;
+ghost bool internal_reverted;
+ghost uint32 function_id;
+
+hook CALL(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/spec/Reentrancy.spec:5]: Add comments. cvlBu example should be documented.

File: CVLByExample/Reentrancy/certora/spec/Reentrancy.spec [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    internal_reverted = lastReverted;
+}
+
+rule no_reentrancy(method f, method g) filtered {f-> !f.isView, g -> !g.isView} {
+    require !called_extcall;
+    require !internal_reverted;
+    env e; calldataarg args;
+    require function_id == g.selector;
+    f@withrevert(e, args);
+    assert called_extcall => internal_reverted;
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/certora/spec/Reentrancy.spec:26]: This should fail? Add in the comment if should fail or pass.

File: CVLByExample/Reentrancy/contracts/VulnerableBankFixed.sol [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        uint256 balance = getUserBalance(msg.sender);
+        require(balance > 0, "Insufficient balance");
+
+        (bool success, ) = msg.sender.call{value: balance}("");
+        require(success, "Failed to send Ether");
+
+        userBalances[msg.sender] = 0;
+    }
+
+    function withdraw(uint256 amount) external noReentrant {
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/contracts/VulnerableBankFixed.sol:26]: Add a comment explaining the failure.

File: CVLByExample/Reentrancy/contracts/VulnerableBankBadFix.sol [hunk 8]
```diff
@@ -0,0 +1,54 @@
+abstract contract ReentrancyGuard {
+    bool internal locked;
+
+    modifier noReentrant() {
+        require(!locked, "No re-entrancy");
+        locked = true;
+        _;
```
Comment (sitvanit83) [inline CVLByExample/Reentrancy/contracts/VulnerableBankBadFix.sol:7]: What is this doing?

Comment (ghilac) [reply CVLByExample/Reentrancy/contracts/VulnerableBankBadFix.sol:7]: this is the reentrancy guard

File: CVLByExample/Reentrancy/contracts/VulnerableBankFixed.sol (hunk 7 above)
Comment (ghilac) [reply CVLByExample/Reentrancy/contracts/VulnerableBankFixed.sol:26]: where?

File: CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf (hunk 3 above)
Comment (ghilac) [reply CVLByExample/Reentrancy/certora/conf/ReentrancyVulnerableBankBadFix.conf:8]: we will use a dedicated version until it is in master

**Review by nd-certora — COMMENTED** [review]
a readme is missing. it should contain :
* what is this exmaple about
* each folder and file
* how to run
* links to run 

also the Readonly is missing that and an explanation of what the builtin rule checks, it can point to the doc and also give a link to the dss lecture

**Review by nd-certora — APPROVED** [review]
few more comments

File: CVLByExample/ReadOnlyReentrancy/README.md [hunk 9]
```diff
@@ -12,4 +12,4 @@ This contract can be checked by running:
 The correct version is `VulnerableBankFixed.sol` the weakness was fixed by moving the update of balances before the external call.
 
 This contract can be verified by running: 
-```certoraRun certora/conf/BuiltinROReentrancyVulnerableBankFixed.conf```
+```certoraRun certora/conf/BuiltinROReentrancyVulnerableBankFixed.conf```
```
Comment (sitvanit83) [inline CVLByExample/ReadOnlyReentrancy/README.md:15]: There is no change here

**Pull request merged**